src/commissioner/ExampleCommissioningStateMachine.h

```diff
@@ -0,0 +1,664 @@
+    {}
+    void Enter() { this->mCtx.Dispatch(Event::Create<NocsrInformation>(mNocsrInformation)); }
+
+private:
+    NocsrInformation mNocsrInformation;
+};
+
+struct SigningCertificates : chip::Commissioner::States::Base<Context>
+{
+    SigningCertificates(Context & ctx, Commissionee & commissionee, NocsrInformation & nocsrInformation, OpCredsIssuer * issuer,
+                        FabricIndex fabricIndex, NodeId nodeId) :
+        Base<Context>(ctx, commissionee, "SigningCertificates"),
+        mNocsrInformation(nocsrInformation), mIssuer(issuer), mFabricIndex(fabricIndex), mNodeId(nodeId)
+    {}
+    void Enter()
+    {
+        // Note that OperationalCredentialsDelegate is just used here for
+        // convenience.  The exmaple app can use any signer interface it
+        // chooses, and from SigningCertificates state context,
+        // everything needed for end-to-end attestation, IPK and admin subject
+        // assignment is accessible.
+        chip::Callback::Callback<chip::Controller::OnNOCChainGeneration> callback(OnDeviceNOCChainGeneration, this);
+        FabricInfo * fabric = this->mCommissionee.mSystemState->Fabrics()->FindFabricWithIndex(mFabricIndex);
+        if (fabric == nullptr)
+        {
+            this->mCtx.Dispatch(Event::Create<Failure>());
+            return;
+        }
+        mIssuer->SetFabricIdForNextNOCRequest(fabric->GetFabricId());
+        mIssuer->SetNodeIdForNextNOCRequest(mNodeId);
+        CHIP_ERROR err =
+            mIssuer->GenerateNOCChain(mNocsrInformation.NocsrElements()->Get(), mNocsrInformation.Signature()->Get(),
+                                      mNocsrInformation.Dac()->Get(), mNocsrInformation.Pai()->Get(), chip::ByteSpan(), &callback);
+        if (err != CHIP_NO_ERROR)
+        {
+            this->mCtx.Dispatch(Event::Create<Failure>());
+        }
+    }
+    void Dispatch(Event event) { this->mCtx.Dispatch(event); }
+
+private:
+    static void OnDeviceNOCChainGeneration(void * context, CHIP_ERROR err, const chip::ByteSpan & derNoc,
+                                           const chip::ByteSpan & derIcac, const chip::ByteSpan & derRcac)
+    {
+
+        ChipLogProgress(Controller, "Received callback from the CA for NOC Chain generation. Status %s", ErrorStr(err));
+        SigningCertificates * state = static_cast<SigningCertificates *>(context);
+        OperationalCredentials opCreds;
+        SuccessOrExit(err);
+
+        {
+            chip::MutableByteSpan chipRcac;
+            SuccessOrExit(err = opCreds.Rcac()->Allocate(chip::Credentials::kMaxCHIPCertLength));
+            chipRcac = opCreds.Rcac()->GetMutable();
+            SuccessOrExit(err = chip::Credentials::ConvertX509CertToChipCert(derRcac, chipRcac));
+            SuccessOrExit(err = opCreds.Rcac()->Set(chipRcac));
+        }
+        if (!derIcac.empty())
+        {
+            chip::MutableByteSpan chipIcac;
```

Comment:
Are these requested in separate states? Can this be handled in those callbacks?

Reply:
This is just an artifact of the use of the `OperationalCredentialsDelegate` in `ExampleCommissioningStateMachine`.  Per https://github.com/project-chip/connectedhomeip/pull/13294/files#r783245705, the value of delegates largely disappears in this design.

In our commissioner code that consumes this, we are able to interact directly with our signing infrastructure in the way we need to without any use of the delegate interfaces.